src/instructlab/sdg/datamixing.py

```diff
@@ -93,16 +101,31 @@ class Recipe:
     prompt used when generating the data in those datasets.
     """
 
-    def __init__(
-        self, initial_datasets: Optional[list] = None, sys_prompt: Optional[str] = ""
-    ):
-        self.recipe = {
-            "datasets": initial_datasets or [],
-            "sys_prompt": sys_prompt,
-        }
-        self.sys_prompt = self.recipe.get("sys_prompt", "")
+    def __init__(self, recipe_path, sys_prompt: Optional[str] = ""):
+        self.recipe_path = recipe_path
+        if recipe_path != "":
+            self.recipe = self._load_recipe()
+        else:
+            self.recipe = {
+                "datasets": [],
+            }
+        if sys_prompt != "":
+            # Overwrite whatever sys_prompt was present in the recipe
+            # with the current one in use in the generation pipeline,
+            # as that's the one that is being used for generation and
+            # thus the one we'll write back out later.
```

Comment:
I guess I'd imagine this being the reverse? The one in the generation pipeline would be the default if there's none specified in the file?

Reply:
Ok, after some back-and-forth discussion, it sounds like we don't really ever need to load the system prompt from these recipe files. It gets written to them to document what was used, but we don't need to load it back in. I'll make that change and unconditionally replace this with the system prompt passed in to the function call.

Reply:
That's pretty bizarre too though ... you load a config file and some of it gets ignored?

Maybe add a 'metadata' section (or similar) to hold `sys_prompt` ... the semantics would be that anything in that section is info written out by the mixer, doesn't influence the mixing process?

Reply:
That's not a bad idea. I'm glad these discussions are happening, although it's starting to become clear this should have happened a couple of weeks ago when these features were first being developed elsewhere.

Reply:
The `sys_prompt` only gets used when writing out the mixed dataset, where we overwrite the system prompt of every source sample to be the prompt specified here. And the reason we include it in the Recipe yaml file is so there's a record of what system prompt exists in the mixed dataset without having to go look inside the dataset file itself. So, maybe when loading a Recipe from disk or when writing it out we should always include the sys_prompt, and take the sys_prompt out of the constructor here. So the constructor would have less logic - loads a recipe from disk or doesn't, but none of this logic about which sys_prompt to use.

Then, within our own code, we'd change the sys_prompt of the recipe after loading it from disk, so that it has the system prompt we're using during generation.

Instead of:
```python
        self.knowledge_recipe = Recipe(
            recipe_path="src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml",
            sys_prompt=self.sys_prompt,
        )
```

we'd move to something like:
```python
        self.knowledge_recipe = Recipe(
            recipe_path="src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml",
        )
        self.knowledge_recipe.sys_prompt = self.sys_prompt
```

This would perhaps be cleaner and make more sense? It removes the overwriting of sys_prompt concerns out of the Recipe class and lets that just focus on loading/dumping itself. And it still lets us override the sys_prompt in our usage where we want to force the system prompt to be what we ran the generation pipeline with, regardless of what was or was not specified in some default recipe.

Reply:
Isn't that the same as just not reading `sys_prompt` from the recipe file at all - e.g. by writing it out in a `metadata` section, and just not reading that section when loading?

Reply:
Resolved by now writing the sys_prompt out to a metadata section.